src/librbd/DeepCopyRequest.cc

```diff
@@ -0,0 +1,383 @@
+    m_snap_id_end(snap_id_end), m_object_number(object_number),
+    m_work_queue(work_queue), m_snap_seqs(snap_seqs), m_prog_ctx(prog_ctx),
+    m_on_finish(on_finish), m_cct(dst_image_ctx->cct),
+    m_lock(unique_lock_name("DeepCopyRequest::m_lock", this)) {
+}
+
+template <typename I>
+DeepCopyRequest<I>::~DeepCopyRequest() {
+  assert(m_snapshot_copy_request == nullptr);
+  assert(m_image_copy_request == nullptr);
+}
+
+template <typename I>
+void DeepCopyRequest<I>::send() {
+  int r = validate_copy_points();
+  if (r < 0) {
+    finish(r);
+    return;
+  }
+
+  send_copy_snapshots();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::cancel() {
+  Mutex::Locker locker(m_lock);
+
+  ldout(m_cct, 20) << dendl;
+
+  m_canceled = true;
+
+  if (m_snapshot_copy_request != nullptr) {
+    m_snapshot_copy_request->cancel();
+  }
+
+  if (m_image_copy_request != nullptr) {
+    m_image_copy_request->cancel();
+  }
+}
+
+template <typename I>
+void DeepCopyRequest<I>::send_copy_snapshots() {
+  m_lock.Lock();
+  if (m_canceled) {
+    m_lock.Unlock();
+    finish(-ECANCELED);
+    return;
+  }
+
+  ldout(m_cct, 20) << dendl;
+
+  Context *ctx = create_context_callback<
+    DeepCopyRequest<I>, &DeepCopyRequest<I>::handle_copy_snapshots>(this);
+  m_snapshot_copy_request = SnapshotCopyRequest<I>::create(
+    m_src_image_ctx, m_dst_image_ctx, m_work_queue, m_snap_seqs, ctx);
+  m_snapshot_copy_request->get();
+  m_lock.Unlock();
+
+  m_snapshot_copy_request->send();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::handle_copy_snapshots(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    m_snapshot_copy_request->put();
+    m_snapshot_copy_request = nullptr;
+    if (r == 0 && m_canceled) {
+      r = -ECANCELED;
+    }
+  }
+
+  if (r == -ECANCELED) {
+    ldout(m_cct, 10) << "snapshot copy canceled" << dendl;
+    finish(r);
+    return;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to copy snapshot metadata: " << cpp_strerror(r)
+                 << dendl;
+    finish(r);
+    return;
+  }
+
+  send_set_size();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::send_set_size() {
+  uint64_t size;
+  {
+    RWLock::RLocker src_locker(m_src_image_ctx->snap_lock);
+    size = m_src_image_ctx->get_image_size(CEPH_NOSNAP);
+  }
+
+  m_dst_image_ctx->owner_lock.get_read();
+  m_dst_image_ctx->snap_lock.get_read();
+  if (m_dst_image_ctx->get_image_size(CEPH_NOSNAP) == size) {
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    send_copy_image();
+    return;
+  }
+
+
+  ldout(m_cct, 20) << size << dendl;
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::set_size(&op, size);
+
+  Context *finish_op_ctx = nullptr;
+  if (m_dst_image_ctx->test_features(RBD_FEATURE_EXCLUSIVE_LOCK,
+                                     m_dst_image_ctx->snap_lock)) {
+    if (m_dst_image_ctx->exclusive_lock == nullptr) {
```

Comment:
If the ```EXCLUSIVE_LOCK``` feature bit is set, I would expect ```ImageCtx::exclusive_lock``` to not be null unless the destination image is currently set to a snapshot. If the exclusive lock is not null, you should ```start_op``` and then fail if ```finish_op_ctx``` is a nullptr.